sys/include/libufs.h

```diff
@@ -10,11 +10,22 @@
 
 
 struct Chan;
-struct mount;
 struct vnode;
+struct thread;
 
+/* Wrapper for a UFS mount.  Should support reading from both kernel and user
+ * space (eventually)
+ */
+struct mount {
+	struct ufsmount*	mnt_data;
+	struct Chan*		chan;
+	int32_t			(*read)(struct mount*, void*, int32_t, int64_t);
```

Comment:
Were it me, I would make it work for one thing, and worry about the other later when we've figured out how it all works. The reason is that this is getting kinda nasty at this point, doing things you don't want to do in kernel for the eventual and maybe never happens user mode case. My belief is we should make this right for kernel first, and then figure out the user case if and when we need it.